sklearn/neighbors/lof.py

```diff
@@ -0,0 +1,317 @@
+        This method allows to generalize prediction to new observations (not
+        in the training set). As LOF originally does not deal with new data,
+        this method is kept private.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features), default=None
+            The query sample or samples to compute the Local Outlier Factor
+            wrt to the training samples. If None, makes prediction on the
+            training data without considering them as their own neighbors.
+
+        Returns
+        -------
+        is_inlier : array of shape (n_samples,)
+            Returns 1 for anomalies/outliers and -1 for inliers.
+        """
+        check_is_fitted(self, ["threshold_", "outlier_factor_",
+                               "_distances_fit_X_",
+                               "_neighbors_indices_fit_X_"])
+
+        if X is not None:
+            X = check_array(X, accept_sparse='csr')
+            is_inlier = np.ones(X.shape[0], dtype=int)
+            is_inlier[self.decision_function(X) <= self.threshold_] = -1
+        else:
+            is_inlier = np.ones(self._fit_X.shape[0], dtype=int)
+            is_inlier[-self.outlier_factor_ <= self.threshold_] = -1
+
+        return is_inlier
+
+    def decision_function(self, X):
+        """Opposite of the Local Outlier Factor of X (as bigger is better).
+
+        The argument X is supposed to contain *new data*: if X contains a
+        point from training, it consider the later in its own neighborhood.
+        Also, the samples in X are not considered in the neighborhood of any
+        point.
+        The decision function on training data is available by considering the
+        opposite of the outlier_factor_ attribute.
+
+        Parameters
+        ----------
+
+        X : array-like, shape (n_samples, n_features)
+            The query sample or samples to compute the Local Outlier Factor
+            w.r.t. the training samples.
+
+        Returns
+        -------
+        lof_scores : array of shape (n_samples,)
+            The Local Outlier Factor of each input samples. The lower,
+            the more abnormal.
+        """
+        check_is_fitted(self, ["threshold_", "outlier_factor_",
+                               "_distances_fit_X_",
+                               "_neighbors_indices_fit_X_"])
+
+        X = check_array(X, accept_sparse='csr')
+
+        distances_X, neighbors_indices_X = self.kneighbors(X)
+        # as bigger is better:
+        return -self._local_outlier_factor(distances_X, neighbors_indices_X)
+
+    def _local_outlier_factor(self, distances_X, neighbors_indices_X):
+        """Compute the local outlier factor (LOF)
+
+        It is the average of the ratio of the local reachability density of
+        a sample and those of its k-nearest neighbors.
+
+        Parameters
+        ----------
+        distances_X : Array of shape (n_query, self.n_neighbors).
```

Comment:
and it should be:

distances_X : array, shape (n_query, self.n_neighbors)

no dot at the end and a , between type and shape